<a href="https://colab.research.google.com/github/daveveed/Geo-Data-Science/blob/main/Processing_data_with_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [38]:
import pandas as pd
import os, requests

In [39]:
data_folder = 'data'
output_folder = 'output'

if not os.path.exists(data_folder):
    os.makedirs(data_folder)
if not os.path.exists(output_folder):
    os.makedirs(output_folder)

In [40]:
def download(url):

  filename = os.path.join(data_folder, os.path.basename(url))

  if not os.path.exists(filename):
    response = requests.get(url, allow_redirects=True)
    with open(filename, 'wb') as f:
      f.write(response.content)
    print('downloaded', filename)

download('https://raw.githubusercontent.com/daveveed/Geo-Data-Science/refs/heads/main/Data/Temp/Rovaniemi%20railway%20station_%201.1.2024%20-%2010.8.2024_temp.csv')

In [41]:
file_path = os.path.join(data_folder, 'Rovaniemi%20railway%20station_%201.1.2024%20-%2010.8.2024_temp.csv')

data = pd.read_csv(file_path)

data.head()

,Observation station,Year,Month,Day,Time [Local time],Air temperature mean [°C]
0,Rovaniemi railway station,2024,1,1,01:00,-24.2
1,Rovaniemi railway station,2024,1,1,02:00,-25.2
2,Rovaniemi railway station,2024,1,1,03:00,-26
3,Rovaniemi railway station,2024,1,1,04:00,-26.9
4,Rovaniemi railway station,2024,1,1,05:00,-27.3


In [42]:
data.columns

Index(['Observation station', 'Year', 'Month', 'Day', 'Time [Local time]',
       'Air temperature mean [°C]'],
      dtype='object')

## renaming columns

In [43]:
new_names = {'Observation station': 'STATION',
             'Time [Local time]': 'TIME',
             'Air temperature mean [°C]': 'TEMP'}

In [44]:
data = data.rename(columns=new_names)
data.columns

Index(['STATION', 'Year', 'Month', 'Day', 'TIME', 'TEMP'], dtype='object')

In [45]:
data = data.rename(columns={
    'TEMP': 'TEMP_C'
})
data.columns

Index(['STATION', 'Year', 'Month', 'Day', 'TIME', 'TEMP_C'], dtype='object')

In [46]:
data.shape

(5327, 6)

In [47]:
data.head()

,STATION,Year,Month,Day,TIME,TEMP_C
0,Rovaniemi railway station,2024,1,1,01:00,-24.2
1,Rovaniemi railway station,2024,1,1,02:00,-25.2
2,Rovaniemi railway station,2024,1,1,03:00,-26
3,Rovaniemi railway station,2024,1,1,04:00,-26.9
4,Rovaniemi railway station,2024,1,1,05:00,-27.3


In [48]:
data.dtypes

,0
STATION,object
Year,int64
Month,int64
Day,int64
TIME,object
TEMP_C,object


In [49]:
data.describe()

,Year,Month,Day
count,5327.0,5327.000000,5327.000000
mean,2024.0,4.172705,15.287028
std,0.0,2.116008,8.882237
min,2024.0,1.000000,1.000000
25%,2024.0,2.000000,7.000000
50%,2024.0,4.000000,15.000000
75%,2024.0,6.000000,23.000000
max,2024.0,8.000000,31.000000


## using functions with pandas

In [50]:
def celsiustofahrenheit(temp_c):
  temp_f = (temp_c * 9/5) + 32
  return temp_f

In [51]:
celsiustofahrenheit(0)

32.0

In [52]:
data['TEMP_C'] = pd.to_numeric(data['TEMP_C'], errors='coerce')
data['TEMP_F'] = data['TEMP_C'].apply(celsiustofahrenheit)
data.head()

,STATION,Year,Month,Day,TIME,TEMP_C,TEMP_F
0,Rovaniemi railway station,2024,1,1,01:00,-24.2,-11.56
1,Rovaniemi railway station,2024,1,1,02:00,-25.2,-13.36
2,Rovaniemi railway station,2024,1,1,03:00,-26.0,-14.80
3,Rovaniemi railway station,2024,1,1,04:00,-26.9,-16.42
4,Rovaniemi railway station,2024,1,1,05:00,-27.3,-17.14


## Parsing dates

In [53]:
data['MONTH_STR'] = data['Month'].astype(str)
data['DAY_STR'] = data['Day'].astype(str)

In [54]:
data['MONTH_DAY'] = data['MONTH_STR'].str.zfill(2) + '-' + data['DAY_STR'].str.zfill(2)
data.head()

,STATION,Year,Month,Day,TIME,TEMP_C,TEMP_F,MONTH_STR,DAY_STR,MONTH_DAY
0,Rovaniemi railway station,2024,1,1,01:00,-24.2,-11.56,1,1,01-01
1,Rovaniemi railway station,2024,1,1,02:00,-25.2,-13.36,1,1,01-01
2,Rovaniemi railway station,2024,1,1,03:00,-26.0,-14.80,1,1,01-01
3,Rovaniemi railway station,2024,1,1,04:00,-26.9,-16.42,1,1,01-01
4,Rovaniemi railway station,2024,1,1,05:00,-27.3,-17.14,1,1,01-01


In [55]:
## pandas.to_datetime()

data['YEAR_STR'] = data['Year'].astype(str)
data['DATE_STR'] = data['YEAR_STR'].str.zfill(2)  + data['MONTH_STR'].str.zfill(2) + data['DAY_STR'].str.zfill(2)
data.head()

,STATION,Year,Month,Day,TIME,TEMP_C,TEMP_F,MONTH_STR,DAY_STR,MONTH_DAY,YEAR_STR,DATE_STR
0,Rovaniemi railway station,2024,1,1,01:00,-24.2,-11.56,1,1,01-01,2024,20240101
1,Rovaniemi railway station,2024,1,1,02:00,-25.2,-13.36,1,1,01-01,2024,20240101
2,Rovaniemi railway station,2024,1,1,03:00,-26.0,-14.80,1,1,01-01,2024,20240101
3,Rovaniemi railway station,2024,1,1,04:00,-26.9,-16.42,1,1,01-01,2024,20240101
4,Rovaniemi railway station,2024,1,1,05:00,-27.3,-17.14,1,1,01-01,2024,20240101


In [56]:
# convert to datetime
data['DATE'] = pd.to_datetime(data['DATE_STR'])
data.head()

,STATION,Year,Month,Day,TIME,TEMP_C,TEMP_F,MONTH_STR,DAY_STR,MONTH_DAY,YEAR_STR,DATE_STR,DATE
0,Rovaniemi railway station,2024,1,1,01:00,-24.2,-11.56,1,1,01-01,2024,20240101,2024-01-01
1,Rovaniemi railway station,2024,1,1,02:00,-25.2,-13.36,1,1,01-01,2024,20240101,2024-01-01
2,Rovaniemi railway station,2024,1,1,03:00,-26.0,-14.80,1,1,01-01,2024,20240101,2024-01-01
3,Rovaniemi railway station,2024,1,1,04:00,-26.9,-16.42,1,1,01-01,2024,20240101,2024-01-01
4,Rovaniemi railway station,2024,1,1,05:00,-27.3,-17.14,1,1,01-01,2024,20240101,2024-01-01


In [57]:
data['DATE'].dt.year.head()

,DATE
0,2024
1,2024
2,2024
3,2024
4,2024


In [58]:
data['DATE'].dt.month.tail()

,DATE
5322,8
5323,8
5324,8
5325,8
5326,8


In [59]:
data['DATE'].dt.day.nunique()

31

For the final analysis, we need combined information of the month and day. Here is another way we could achieve this:

In [60]:
# Extract month and day from the full date (as a string in 'MM-DD' format)
data["MONTH_DAY_DT"] = data["DATE"].dt.strftime("%m-%d")

When using pd.to_datetime() with only month (%m) and day (%d) in the format string, pandas defaults to using the year 1900. This behavior occurs because the datetime module requires a full date, including a year, to function properly. Without a specified year, pandas fills in the year as 1900.

If you need to represent just the month and day without the year, like we do here, we can treat the values as strings using dt.strftime("%m-%d"). This way, we avoid the default year issue.

Alternatively, if you plan to perform date-based operations, you should retain the full date and simply ignore the year in your calculations.

In [61]:
data["MONTH_DAY_DT"]

,MONTH_DAY_DT
0,01-01
1,01-01
2,01-01
3,01-01
4,01-01
...,...
5322,08-09
5323,08-09
5324,08-09
5325,08-09


## Aggregating data in pandas by grouping

In [62]:
len(data)

5327

In [63]:
data.shape

(5327, 14)

In [64]:
data.head()

,STATION,Year,Month,Day,TIME,TEMP_C,TEMP_F,MONTH_STR,DAY_STR,MONTH_DAY,YEAR_STR,DATE_STR,DATE,MONTH_DAY_DT
0,Rovaniemi railway station,2024,1,1,01:00,-24.2,-11.56,1,1,01-01,2024,20240101,2024-01-01,01-01
1,Rovaniemi railway station,2024,1,1,02:00,-25.2,-13.36,1,1,01-01,2024,20240101,2024-01-01,01-01
2,Rovaniemi railway station,2024,1,1,03:00,-26.0,-14.80,1,1,01-01,2024,20240101,2024-01-01,01-01
3,Rovaniemi railway station,2024,1,1,04:00,-26.9,-16.42,1,1,01-01,2024,20240101,2024-01-01,01-01
4,Rovaniemi railway station,2024,1,1,05:00,-27.3,-17.14,1,1,01-01,2024,20240101,2024-01-01,01-01


In [67]:
grouped = data.groupby('MONTH_DAY')
type(grouped)

pandas.core.groupby.generic.DataFrameGroupBy

In [68]:
len(grouped)

223

In [69]:
data['MONTH_DAY'].nunique()

223

Length of the grouped object should be the same as the number of unique values in the column we used for grouping. For each unique value, there is a group of data. This makes sense, as we have 223 days from first of January to the end of August.

In [70]:
# We can check the “names” of each group.
grouped.groups.keys()

dict_keys(['01-01', '01-02', '01-03', '01-04', '01-05', '01-06', '01-07', '01-08', '01-09', '01-10', '01-11', '01-12', '01-13', '01-14', '01-15', '01-16', '01-17', '01-18', '01-19', '01-20', '01-21', '01-22', '01-23', '01-24', '01-25', '01-26', '01-27', '01-28', '01-29', '01-30', '01-31', '02-01', '02-02', '02-03', '02-04', '02-05', '02-06', '02-07', '02-08', '02-09', '02-10', '02-11', '02-12', '02-13', '02-14', '02-15', '02-16', '02-17', '02-18', '02-19', '02-20', '02-21', '02-22', '02-23', '02-24', '02-25', '02-26', '02-27', '02-28', '02-29', '03-01', '03-02', '03-03', '03-04', '03-05', '03-06', '03-07', '03-08', '03-09', '03-10', '03-11', '03-12', '03-13', '03-14', '03-15', '03-16', '03-17', '03-18', '03-19', '03-20', '03-21', '03-22', '03-23', '03-24', '03-25', '03-26', '03-27', '03-28', '03-29', '03-30', '03-31', '04-01', '04-02', '04-03', '04-04', '04-05', '04-06', '04-07', '04-08', '04-09', '04-10', '04-11', '04-12', '04-13', '04-14', '04-15', '04-16', '04-17', '04-18', '04-19',

accessing data for one group

In [71]:
day = '07-17'
group1 = grouped.get_group(day)

In [74]:
group1.head()

,STATION,Year,Month,Day,TIME,TEMP_C,TEMP_F,MONTH_STR,DAY_STR,MONTH_DAY,YEAR_STR,DATE_STR,DATE,MONTH_DAY_DT
4750,Rovaniemi railway station,2024,7,17,00:00,17.7,63.86,7,17,07-17,2024,20240717,2024-07-17,07-17
4751,Rovaniemi railway station,2024,7,17,01:00,16.4,61.52,7,17,07-17,2024,20240717,2024-07-17,07-17
4752,Rovaniemi railway station,2024,7,17,02:00,15.3,59.54,7,17,07-17,2024,20240717,2024-07-17,07-17
4753,Rovaniemi railway station,2024,7,17,03:00,14.5,58.10,7,17,07-17,2024,20240717,2024-07-17,07-17
4754,Rovaniemi railway station,2024,7,17,04:00,14.3,57.74,7,17,07-17,2024,20240717,2024-07-17,07-17


So, as noted above, one group is a pandas DataFrame! This is really useful, because we can now use all the familiar DataFrame methods for calculating statistics, etc. for this specific group. We can, for example, calculate the average values for all variables using the statistical functions that we have seen already (e.g. mean, std, min, max, median, etc.).

Thus, we can get an average temperature for a given day using the .mean() function that we already did during Lesson 5. Let’s calculate the mean for following attributes all at once:

In [75]:
mean_values = group1[['TEMP_C', 'TEMP_F']].mean()
mean_values

,0
TEMP_C,18.879167
TEMP_F,65.982500


In [76]:
# Create an empty DataFrame for the aggregated values
daily_data = pd.DataFrame()

# The columns that we want to aggregate
mean_cols = ["TEMP_F", "TEMP_C"]

# Iterate over the groups
for key, group in grouped:

    # Calculate mean
    mean_values = group[mean_cols].mean()

    # Add the ´key´ (i.e. the date+time information) into the aggregated values
    mean_values["MONTH_DAY"] = key

    # Convert the mean_values series to a DataFrame and make it have a row orientation
    row = mean_values.to_frame().transpose()

    # Concatenate the aggregated values into the daily_data DataFrame
    daily_data = pd.concat([daily_data, row], ignore_index=True)

In [77]:
daily_data

,TEMP_F,TEMP_C,MONTH_DAY
0,-16.693913,-27.052174,01-01
1,-19.7275,-28.7375,01-02
2,-22.5625,-30.3125,01-03
3,-19.795,-28.775,01-04
4,-8.3125,-22.395833,01-05
...,...,...,...
218,64.685,18.158333,08-06
219,62.9525,17.195833,08-07
220,63.26,17.366667,08-08
221,66.17,18.983333,08-09


## Detecting warm days in July

In [78]:
data.head(2)

,STATION,Year,Month,Day,TIME,TEMP_C,TEMP_F,MONTH_STR,DAY_STR,MONTH_DAY,YEAR_STR,DATE_STR,DATE,MONTH_DAY_DT
0,Rovaniemi railway station,2024,1,1,01:00,-24.2,-11.56,1,1,01-01,2024,20240101,2024-01-01,01-01
1,Rovaniemi railway station,2024,1,1,02:00,-25.2,-13.36,1,1,01-01,2024,20240101,2024-01-01,01-01


In [79]:
july_days = data[data['Month'] == 7]
july_days = july_days[['TEMP_F', 'TEMP_C', 'MONTH_DAY']]

In [80]:
july_days.head()

,TEMP_F,TEMP_C,MONTH_DAY
4366,60.80,16.0,07-01
4367,59.54,15.3,07-01
4368,58.46,14.7,07-01
4369,57.74,14.3,07-01
4370,57.38,14.1,07-01


In [81]:
grouped = july_days.groupby('MONTH_DAY')

In [82]:
daily_mean_july = grouped.mean()
daily_mean_july.head()

,TEMP_F,TEMP_C
MONTH_DAY,,
07-01,59.8025,15.445833
07-02,59.2175,15.120833
07-03,59.0525,15.029167
07-04,63.4625,17.479167
07-05,58.6175,14.787500


Finally, we can sort and check the highest temperature values. We can sort the data frame in a descending order to do this.

In [84]:
daily_mean_july.sort_values(by="TEMP_C", ascending=False).head(10)

,TEMP_F,TEMP_C
MONTH_DAY,,
07-26,73.1225,22.845833
07-24,73.0100,22.783333
07-25,72.3050,22.391667
07-23,71.4275,21.904167
07-27,70.7075,21.504167
07-22,70.1300,21.183333
07-20,68.7425,20.412500
07-18,68.1425,20.079167
07-21,68.1200,20.066667


calculating the maximum recorded temperature for each day

In [85]:
data.head()

,STATION,Year,Month,Day,TIME,TEMP_C,TEMP_F,MONTH_STR,DAY_STR,MONTH_DAY,YEAR_STR,DATE_STR,DATE,MONTH_DAY_DT
0,Rovaniemi railway station,2024,1,1,01:00,-24.2,-11.56,1,1,01-01,2024,20240101,2024-01-01,01-01
1,Rovaniemi railway station,2024,1,1,02:00,-25.2,-13.36,1,1,01-01,2024,20240101,2024-01-01,01-01
2,Rovaniemi railway station,2024,1,1,03:00,-26.0,-14.80,1,1,01-01,2024,20240101,2024-01-01,01-01
3,Rovaniemi railway station,2024,1,1,04:00,-26.9,-16.42,1,1,01-01,2024,20240101,2024-01-01,01-01
4,Rovaniemi railway station,2024,1,1,05:00,-27.3,-17.14,1,1,01-01,2024,20240101,2024-01-01,01-01


In [86]:
summer_days = data[['TEMP_C', 'MONTH_DAY']]
grouped = summer_days.groupby('MONTH_DAY')
daily_max = grouped.max()
daily_max = daily_max.rename(columns={'TEMP_C': 'MAX_TEMP_C'})
daily_max.head()

,MAX_TEMP_C
MONTH_DAY,
01-01,-24.2
01-02,-25.7
01-03,-28.5
01-04,-25.0
01-05,-19.5


In [89]:
grouped.head()

,TEMP_C,MONTH_DAY
0,-24.2,01-01
1,-25.2,01-01
2,-26.0,01-01
3,-26.9,01-01
4,-27.3,01-01
...,...,...
5303,15.8,08-09
5304,15.0,08-09
5305,13.2,08-09
5306,12.2,08-09
